[Shopify] Connector - Sales Channels publishing for new products

petemchlk opened the pull request:
### Sales Channels publishing for new products

This PR contains new functionality where user can choose Sales Channels which will new products be published to on creation from BC.

Solution contains new page and table - Shopify Sales Channels available from Shopify Shop. On page user can import sales channels available for shop and choose to which new products should be published to. If user does not choose any of the channels, then channel "Online Shop" is being used by default.

In case user did not import Sales Channels before running create product sales channels are pulled from Shopify on before publishing and default channel is used.

Pulled sales channels are updated in BC - in case the channel was removed in Shopify its removed from list in BC.

### Modified/new objects

- New GQL query codeunit was created to import the Sales Channels.
- New GQL Query Enum Value 
- New codeunit Shpfy Sales Channel API to handle sales channels import
- Removed detracted published = true from Shpfy Product API, added procedures to handle new way of publishing new products to sales channels
- Sales Channels page and table
- New action on Shopify Shop Card page - Sales Channels

Commit:
add sales channels import

Commit:
add publishing process on create, modify structure to be able to choose channel

Commit:
code refactor

Commit:
Create import test

Commit:
add tests, refactor code

Commit:
add tests

Commit:
remove empty line

Commit:
udpate gql query, add default field and default channel publishing in process

Commit:
fix tests, add test for trgering publish from create prod

tinestaric commented on Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductAPI.Codeunit.al:
```diff
@@ -567,4 +569,63 @@ codeunit 30176 "Shpfy Product API"
+        exit(true);
+    end;
+
+    local procedure CreateProductPublishGraphQuery(ShopifyProduct: Record "Shpfy Product"; var SalesChannel: Record "Shpfy Sales Channel"; GraphQuery: TextBuilder)
```
nit: GraphQuery could be text, or maybe a return variable of text

Passing it as text builder implies that there's more going to be added.

tinestaric commented on Apps/W1/Shopify/app/src/Products/Codeunits/ShpfySalesChannelAPI.Codeunit.al:
```diff
@@ -0,0 +1,95 @@
+        SalesChannel: Record "Shpfy Sales Channel";
+        ChannelId: BigInteger;
+    begin
+        if CurrentChannels.Count() <> 0 then
```
why this condition? Doesn't the foreach just skip if CurrentChannels is empty?

tinestaric commented on Apps/W1/Shopify/app/src/Products/Codeunits/ShpfySalesChannelAPI.Codeunit.al:
```diff
@@ -0,0 +1,95 @@
+                SalesChannel."Shop Code" := ShopCode;
+                SalesChannel.Insert(true);
+            end else
+                CurrentChannels.Remove(ChannelId);
```
shouldn't the CurrentChannels.Remove be called also in the case where SalesChannel gets created?

Otherwise, the Channel doesn't exist yet -> we create it -> we don't remove it from the CurrentChannels -> RemoveExisting deletes it becose the id is still in the list?

tinestaric replied:
Ahh, it's never in the CurrentChannels to begin with, because it doesn't exist... *facepalm*

tinestaric commented on Apps/W1/Shopify/test/Products/ShpfySalesChannelTest.Codeunit.al:
```diff
@@ -0,0 +1,274 @@
+        // [WHEN] Invoking the procedure: ShopifyProductAPI.PublishProduct(ShopifyProduct)
+        ShopifyProductAPI.PublishProduct(ShopifyProduct);
+
+        // [THEN] Procedure exits without publishing the product.
```
how do you assert that?

tinestaric commented on Apps/W1/Shopify/test/Products/ShpfySalesChannelTest.Codeunit.al:
```diff
@@ -0,0 +1,274 @@
+        // [WHEN] Invoking the procedure: ShopifyProductAPI.PublishProduct(ShopifyProduct)
+        ShopifyProductAPI.PublishProduct(ShopifyProduct);
+
+        // [THEN] Procedure exits without publishing the product.
```
same here.

Commit:
PR fixes

tinestaric commented on Apps/W1/Shopify/app/src/Products/Codeunits/ShpfySalesChannelAPI.Codeunit.al:
```diff
@@ -0,0 +1,94 @@
+        Channels: List of [BigInteger];
+    begin
+        SalesChannel.SetRange("Shop Code", ShopCode);
+        if SalesChannel.FindSet() then
```
One best practice that I have yet to get used to, but I have seen MS adding it on our pull requests before the merge it in is the use of SetLoadFields and ReadIsolation, so I think we need to start getting used to using it

Commit:
update object ids, premission sets, gql type enum ids fixed

Commit:
fix summaries

Commit:
create get next sales channels

Commit:
fix function name

Commit:
saleschannels pagination

Commit:
adjust tests to sales channels paging

Commit:
refactor

Commit:
fix tooltip

tinestaric committed:
[Shopify] Customer/Company Metafields (#27542)

This pull request does not have a related issue as it's part of delivery
for development agreed directly with @AndreiPanko
Fixes #26819

### Existing metafield engine applied to integrate Customer and Company
metafields

Changes done to apply the engine:

- Extended `ShpfyGQLCompany.Codeunit.al` and
`ShpfyGQLCustomer.Codeunit.al` GQL query to get metafields with required
properties in response. Removed the namespace from query to retreive all
metafields.
- Created 2 new GraphQL queries `Shpfy GQL CustomerMetafieldIds` and
`Shpfy GQL CompanyMetafieldIds` to make `RetrieveMetafieldsFromShopify`
function work with Company and Customer.
- Created 2 new IMetafield Owner Type implementation codeunits for
Company and Customer
- Added values to related interface enums
- Added UpdateMetafield functions in API codeunits to update metafields
if there are metafields in response
- Removed old not working code to update metafields
- Add Metafields actions to Company/Customer list and card pages.
- Add owner type Company to `GetOwnerType` procedure in `Shpfy
Metafield` table
- Changed parameters in `CollectMetafieldIds` local procedure in `Shpfy
Metafield API` codeunit to support all Ownet types in function

Fixes
[AB#556298](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/556298)

---------

Co-authored-by: Piotr Michalak <piotr.michalak@companial.com>

tinestaric committed:
[Shopify] - Add Item As Variant Tests (#27541)

This pull request does not have a related issue as it's part of delivery
for development agreed directly with @AndreiPanko
Fixes #26819

### Add Item As Variant Test

Automated tests for changes #26712 


Fixes
[AB#555979](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/555979)

---------

Co-authored-by: Piotr Michalak <piotr.michalak@companial.com>

aholstrup1 committed:
Syncing with version 26.0.26183.0 (#27582)

JesperSchulz committed:
Merge branch 'main' into dev/pmi/SalesChannels

Commit:
PR fixes

Commit:
Merge branch 'dev/pmi/SalesChannels' of https://github.com/GediminasGaubys/ALAppExtensions into dev/pmi/SalesChannels

glachana committed:
[Shopify] Fix sync catalog price issue (#27555)

<!-- Thank you for submitting a Pull Request. If you're new to
contributing to AlAppExtensions please read our pull request guideline
below
* https://github.com/microsoft/ALAppExtensions/blob/main/CONTRIBUTING.md
-->
#### Summary <!-- Provide a general summary of your changes -->
When users try to synchronize prices through the action available on
Shopify Catalogs page. Only the 100 first prices are synchronized. So if
the Shopify Catalog contains more than 100 product, only 100 of them
will be updated.

This problem results from a difference between two GraphQL requests made
during this process:

1) codeunit 30296 "Shpfy GQL CatalogPrices"
'{"query": "query { catalog(id: "gid://shopify/Catalog/{{CatalogId}}") {
id priceList {id currency prices(first:100) {edges {cursor node {variant
{id product {id}} price {amount} compareAtPrice {amount}}} pageInfo
{hasNextPage}}}}}"}'

2) codeunit 30297 "Shpfy GQL NextCatalogPrices"
'{"query": "query { catalog(id: "gid://shopify/Catalog/{{CatalogId}}") {
id priceList {id prices(first:100, after:"{{After}}") {edges {cursor
node {variant {id} price {amount} compareAtPrice {amount}}} pageInfo
{hasNextPage}}}}}"}'

Some necessary nodes are missing in NextCatalogPrices. Here is a fixed
request :

'{"query": "query { catalog(id: "gid://shopify/Catalog/{{CatalogId}}") {
id priceList {id currency prices(first:100, after:"{{After}}") {edges
{cursor node {variant {id product {id}} price {amount} compareAtPrice
{amount}}} pageInfo {hasNextPage}}}}}"}'


#### Work Item(s) <!-- Add the issue number here after the #. The issue
needs to be open and approved. Submitting PRs with no linked issues or
unapproved issues is highly discouraged. -->
Fixes #27535


Fixes
[AB#556528](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/556528)

onbuyuka committed:
Update Apps/W1/Shopify/app/src/Products/Tables/ShpfySalesChannel.Table.al

GediminasGaubys committed:
[Shopify] Shopify Return Location on Refunds tests (#27578)

This pull request does not have a related issue as it's part of delivery
for development agreed directly with @AndreiPanko
Fixes https://github.com/microsoft/ALAppExtensions/issues/26819

Add Item As Variant Test
Automated tests for changes
https://github.com/microsoft/ALAppExtensions/pull/26728
Fixes
[AB#556527](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/556527)

---------

Co-authored-by: Piotr Michalak <piotr.michalak@companial.com>

GediminasGaubys committed:
[Shopify] - Export Posted Sales Invoices to Shopify as Orders Tests (#27570)

This pull request does not have a related issue as it's part of delivery
for development agreed directly with @AndreiPanko

Add Item As Variant Test
Automated tests for changes
https://github.com/microsoft/ALAppExtensions/pull/26748
Fixes #26819
Fixes
[AB#556529](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/556529)

---------

Co-authored-by: Jesper Schulz-Wedde <JesperSchulz@users.noreply.github.com>

aholstrup1 committed:
Reenable test execution (#27587)

<!-- Thank you for submitting a Pull Request. If you're new to
contributing to AlAppExtensions please read our pull request guideline
below
* https://github.com/microsoft/ALAppExtensions/blob/main/CONTRIBUTING.md
-->
#### Summary <!-- Provide a general summary of your changes -->
Reenable tests by uninstalling apps that aren't supported in W1 before
we start running tests. This will ensure tests don't accidently trigger
e.g. event subscribers from apps that aren't supported in W1 anyway.
Right now we are building, publishing and testing apps on top of a W1
Base App even though some of them aren't supported in W1.

We'll continue to compile the apps. We just won't have them installed
when running tests. With this change, tests are working again:
https://github.com/microsoft/ALAppExtensions/actions/runs/11664233820

Another way we could solve this from AL-Go would could define anti
patterns for app/test folders. In that case we would have the following
in the AL-Go settings `testFolders: [ Apps/W1/*/app/,
!Apps/W1/INTaxEngine/app ] `

#### Work Item(s) <!-- Add the issue number here after the #. The issue
needs to be open and approved. Submitting PRs with no linked issues or
unapproved issues is highly discouraged. -->
Fixes
[AB#556300](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/556300)

blauerPCL committed:
Initial upload of the Dynamics SL Historical Data extension (#27527)

<!-- Thank you for submitting a Pull Request. If you're new to
contributing to AlAppExtensions please read our pull request guideline
below
* https://github.com/microsoft/ALAppExtensions/blob/main/CONTRIBUTING.md
-->
#### Summary: Initial upload of the Dynamics SL Historical Data
extension

#### Work Item(s) <!-- Add the issue number here after the #. The issue
needs to be open and approved. Submitting PRs with no linked issues or
unapproved issues is highly discouraged. -->
Fixes #27620

Fixes
[AB#556923](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/556923)

Groenbech96 committed:
Syncing with version 26.0.26946.0 (#27653)

<!-- Thank you for submitting a Pull Request. If you're new to
contributing to AlAppExtensions please read our pull request guideline
below
* https://github.com/microsoft/ALAppExtensions/blob/main/CONTRIBUTING.md
-->
#### Summary <!-- Provide a general summary of your changes -->

#### Work Item(s) <!-- Add the issue number here after the #. The issue
needs to be open and approved. Submitting PRs with no linked issues or
unapproved issues is highly discouraged. -->
Fixes #

Co-authored-by: magnushar <magnushar@microsoft.com>

Groenbech96 committed:
Syncing with version 26.0.26951.0 (#27654)

<!-- Thank you for submitting a Pull Request. If you're new to
contributing to AlAppExtensions please read our pull request guideline
below
* https://github.com/microsoft/ALAppExtensions/blob/main/CONTRIBUTING.md
-->
#### Summary <!-- Provide a general summary of your changes -->

#### Work Item(s) <!-- Add the issue number here after the #. The issue
needs to be open and approved. Submitting PRs with no linked issues or
unapproved issues is highly discouraged. -->
Fixes #

Co-authored-by: magnushar <magnushar@microsoft.com>

Groenbech96 committed:
Syncing with version 26.0.27052.0 (#27662)

Co-authored-by: Magnus Hartvig Grønbech <magnushar@microsoft.com>

JesperSchulz committed:
Merge branch 'main' into dev/pmi/SalesChannels

JesperSchulz committed:
Fix ID clashes

JesperSchulz committed:
Update IDs
